### PR TITLE
Feat/improve validator messages

### PR DESCRIPTION
**Pull request by slickroot**
# DAH-1457: Improve Validator Messages and Add Comprehensive Test Coverage

## Summary

This PR implements a new modular pipeline-based validation architecture to replace the monolithic `create_task_old` function. The new pipeline uses structured messages with actionable remediation and has comprehensive test coverage for all validation checks.

### 1. Structured Message System
- Centralized all validation messages into message templates with consistent fields
- Added `ValidationEvent` model with structured fields: `event`, `reason_code`, `severity`, `category`, `impact`, `remediation`, `what_we_saw`
- Type-safe reason codes using `Msg.*.reason` pattern
- Actionable remediation guidance for every failure scenario

### 2. Pipeline-Based Validation Architecture
- Introduced modular check-based pipeline replacing monolithic `create_task_old` function
- Each validation concern is now an isolated, testable check with single responsibility
- Checks source dependencies from typed `Context` (services, config, state)
- Pipeline orchestrates check execution with fatal failure and halt support
- Added `SSHCommandRunner` with retry logic and observability for SSH operations
- Structured `CheckResult` model with updates, halt flag, and validation events
- Context immutability: checks return updates instead of mutating shared state
- Clear separation between preparation, validation, and finalization phases

### 3. Comprehensive Test Coverage
Added tests for all 12 previously untested checks, organized in 4 phases:

**Phase 1: Simple Checks (No Mocking)**
- `GpuFingerprintCheck` - GPU UUID change detection
- `GpuUsageCheck` - GPU utilization validation
- `NvmlDigestCheck` - NVIDIA driver integrity

**Phase 2: Mock Services**
- `CapabilityCheck` - GPU capability verification
- `PortConnectivityCheck` - Docker port mapping validation
- `VerifyXCheck` - Hardware validation via VerifyX

**Phase 3: SSH Operations**
- `MachineSpecScrapeCheck` - Machine spec scraping and decryption
- `StartGPUMonitorCheck` - GPU monitor startup
- `UploadFilesCheck` - SFTP file upload

**Phase 4: Complex Checks**
- `FinalizeCheck` - Pipeline result aggregation
- `ScoreCheck` - Score calculation
- `TenantEnforcementCheck` - Rented machine validation flow

### 4. Refactored Existing Tests
- Updated all existing tests to use `Msg.*.reason` for type-safe reason codes
- Standardized mock patterns across all tests
- Improved test helper functions in `tests/helpers.py`

### Flow Diagram

```mermaid
flowchart TD
    Start([Pipeline Start]) --> PrepPhase[Preparation Phase]

    PrepPhase --> GPU_Monitor[StartGPUMonitorCheck]
    GPU_Monitor -->|pass| Upload[UploadFilesCheck]
    Upload -->|pass| Scrape[MachineSpecScrapeCheck]
    Scrape -->|fail/fatal| Fatal1[❌ Fatal Failure]
    Scrape -->|pass| ValidationPhase[Validation Phase]

    ValidationPhase --> GpuCount[GpuCountCheck]
    GpuCount -->|pass| GpuModel[GpuModelValidCheck]
    GpuModel -->|pass| NvmlDigest[NvmlDigestCheck]
    NvmlDigest -->|pass| SpecChange[SpecChangeCheck]
    SpecChange -->|pass| Fingerprint[GpuFingerprintCheck]
    Fingerprint -->|pass| Banned[BannedGpuCheck]
    Banned -->|pass| Duplicate[DuplicateExecutorCheck]
    Duplicate -->|pass| Collateral[CollateralCheck]

    GpuCount -->|fail/fatal| Fatal2[❌ Fatal Failure]
    GpuModel -->|fail/fatal| Fatal2
    NvmlDigest -->|fail/fatal| Fatal2
    SpecChange -->|fail/fatal| Fatal2
    Fingerprint -->|fail/fatal| Fatal2
    Banned -->|fail/fatal| Fatal2
    Duplicate -->|fail/fatal| Fatal2
    Collateral -->|fail/fatal| Fatal2

    Collateral -->|pass| RentalBranch{TenantEnforcementCheck}

    RentalBranch -->|not rented| UnrentedPath[Unrented Path]
    RentalBranch -->|rented & pod not running| Fatal3[❌ Fatal Failure]
    RentalBranch -->|rented & GPU outside tenant| Fail1[❌ Validation Failed]
    RentalBranch -->|rented & valid| RentedFinalize[🎯 Compute Rented Score]

    RentedFinalize --> RentedHalt[✅ Halt Pipeline - Success]

    UnrentedPath --> GpuUsage[GpuUsageCheck]
    GpuUsage -->|fail| Fail2[❌ Validation Failed]
    GpuUsage -->|pass| PortConn[PortConnectivityCheck]
    PortConn -->|fail| Fail3[❌ Validation Failed]
    PortConn -->|pass| VerifyX[VerifyXCheck]
    VerifyX -->|fail| Fail4[❌ Validation Failed]
    VerifyX -->|pass| Capability[CapabilityCheck]
    Capability -->|fail| Fail5[❌ Validation Failed]
    Capability -->|pass| PortCount[PortCountCheck]
    PortCount -->|pass| Score[ScoreCheck]
    Score -->|pass| Finalize[FinalizeCheck]
    Finalize --> Success[✅ Pipeline Complete]

    style Start fill:#e1f5e1
    style Fatal1 fill:#ffcccc
    style Fatal2 fill:#ffcccc
    style Fatal3 fill:#ffcccc
    style Fail1 fill:#ffe6cc
    style Fail2 fill:#ffe6cc
    style Fail3 fill:#ffe6cc
    style Fail4 fill:#ffe6cc
    style Fail5 fill:#ffe6cc
    style RentedHalt fill:#ccffcc
    style Success fill:#ccffcc
    style RentalBranch fill:#fff3cd
    style RentedFinalize fill:#d1ecf1
```

#### 1. Preparation Phase
Sets up the execution environment and gathers machine specifications.

| Check | Fatal? | Purpose |
|-------|--------|---------|
| StartGPUMonitorCheck | No | Starts GPU monitoring agent for live telemetry |
| UploadFilesCheck | No | Uploads validation scripts to executor |
| MachineSpecScrapeCheck | **Yes** | Scrapes and decrypts hardware specifications |

#### 2. Validation Phase
Validates GPU hardware, fingerprints, and policy compliance.

| Check | Fatal? | Purpose |
|-------|--------|---------|
| GpuCountCheck | **Yes** | Validates GPU count is within policy limits |
| GpuModelValidCheck | **Yes** | Validates GPU model is recognized and supported |
| NvmlDigestCheck | **Yes** | Validates NVIDIA driver integrity via MD5 checksum |
| SpecChangeCheck | **Yes** | Detects GPU inventory changes since last validation |
| GpuFingerprintCheck | **Yes** | Detects GPU UUID changes (hardware swaps) |
| BannedGpuCheck | **Yes** | Checks if GPU UUIDs are on banned list |
| DuplicateExecutorCheck | **Yes** | Detects duplicate executor registrations |
| CollateralCheck | No | Validates collateral deposit status |

#### 3. Rental Branching Point
**TenantEnforcementCheck** determines execution path:

**Rented Path (Halts Pipeline)**
1. Check if container is running
2. Verify GPU ownership by tenant container
3. Compute rental score
4. **Halt with success** ✅

**Unrented Path (Continues)**
Proceeds to capability validation checks below.

#### 4. Capability Validation Phase (Unrented Only)

| Check | Fatal? | Purpose |
|-------|--------|---------|
| GpuUsageCheck | No | Validates GPU is idle (not busy with other workloads) |
| PortConnectivityCheck | No | Validates Docker port mapping and connectivity |
| VerifyXCheck | No | Validates hardware specs via VerifyX service |
| CapabilityCheck | No | Validates GPU visibility in Docker containers |
| PortCountCheck | No | Validates sufficient ports are available |

#### 5. Finalization Phase (Unrented Only)

| Check | Fatal? | Purpose |
|-------|--------|---------|
| ScoreCheck | No | Computes actual and job scores based on all validation results |
| FinalizeCheck | No | Aggregates results into structured log and sets success flag |

### Key Behaviors

**Fatal vs Non-Fatal Checks**
- **Fatal checks**: Pipeline stops immediately on failure, returns error
- **Non-fatal checks**: Pipeline continues, may affect score but doesn't stop execution

**Halt Mechanism**
- Only **TenantEnforcementCheck** can halt with success
- Halting prevents remaining checks from running
- Used to short-circuit when machine is already rented

**Context Updates**
Each check can update the pipeline context with state, scores, rental status, verification flags, etc.

## Review this PR

Ways to review this PR is to check out the code and run some synthetic jobs against your executors using the CLI.

## Testing Approach

All tests follow consistent patterns:
- `@pytest.mark.parametrize` for comprehensive scenario coverage
- Mock services with dummy classes that track calls
- `Msg.*.reason` for type-safe reason codes
- Verify both results and service interactions
- Test helper functions for building test contexts

image

## Breaking Changes

None. All changes are backward compatible.

## Future Improvements

While this PR significantly improves the validator architecture, there are areas that could benefit from further refinement:

### 1. Context Object Size
The `Context` object has grown quite large with many fields (score, job_score, collateral_deposited, rented, ssh_pub_keys, etc.). Consider:
- Grouping related fields into nested structures (e.g., `scoring`, `rental_info`)
- Using separate context types for different pipeline phases
- Applying the builder pattern to construct context incrementally

### 2. TenantEnforcementCheck Complexity
This check handles multiple responsibilities (rental detection, container health, GPU ownership, score calculation). Future work could:
- Split into smaller, focused checks (RentalStatusCheck, ContainerHealthCheck, etc.)
- Consider separate pipeline branches for rented vs unrented paths
- Extract score calculation logic to reduce check coupling

### 3. Test Maintenance
Mock service setup is verbose and repetitive across test files. Improvements:
- Create reusable mock factories with sensible defaults
- Add fixture composition for common mock combinations
- Consider property-based testing for edge cases

## Related Issues

- DAH-1457

### CODE REVIEW

File: neurons/validators/src/services/task/checks/port_count.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+    async def run(self, ctx: Context) -> CheckResult:
+        port_mapping = ctx.services.port_mapping
+        redis_service = ctx.services.redis
+        executor_uuid = ctx.executor.uuid
+
+        try:
+            port_count = await port_mapping.get_successful_ports_count(executor_uuid)
+        except Exception:
+            port_count = 0
```
[the remark this answers is not in the document]
Comment (arhangel66) [reply neurons/validators/src/services/task/checks/port_count.py]: The logic is fine for me — we have a fallback to Redis right below.
I mean, we use Redis if the DB isn’t working, or if the DB is empty, or if the port_count from the DB is less than MIN_PORT_COUNT.
